### PR TITLE
Update documentation

### PR DESCRIPTION
**Pull request by anvacaru**
Created [./VERIFICATION.md]() with instructions on how to get started with kevm proofs.
Updated [./include/kframework/foundry.md]() and [./README.md]().

### CODE REVIEW

**Comment by ehildenb** [comment]
@anvacaru make sure to address this one too: https://github.com/runtimeverification/evm-semantics/issues/1536

**Comment by ehildenb** [comment]
Another link to include (EVM opcodes reference): https://www.evm.codes/?fork=merge

Also, linking to solidity docs on ABI encoding would be good: https://docs.soliditylang.org/en/v0.8.19/abi-spec.html

**Comment by ehildenb** [comment]
@anvacaru can you also include this example? https://github.com/runtimeverification/evm-semantics/pull/1578#issue-1595898037

I think we should have a separate file called `verification.md` in the root of the repo, and link to it from the `README.md` towards the top, saying: "If you want to use KEVM to do verification, after installing, follow the instructions here: LINK TO VERIFICATION".

The file should document:

- How to use normal `kevm prove ...` to do verification, and the `--debugger` option.
- How to use `kevm prove --pyk ...` to do verification (as documented in the PR above), and the various options supported there.
- How to use `kevm view-kcfg ...` to view the proof generated with ``kevm prove --pyk ...`, and a GIF demonstrating it.
- A link to `foundry.md` for the more detailed documentation about how to do foundry proofs.

File: include/kframework/foundry.md [hunk 1]
```diff
@@ -1,71 +1,86 @@
...
+   - `—node` - can be a repeated option, display more information about a given node   hash,
+   - `—node-delta` - displays the delta between two given nodes in the KCFG.
+
+- `kevm foundry-view-kcfg`: Launch the more interactive exploration of the KCFG (can be done while exploration is running, must Ctrl-C + relaunch to view updates to KCFG).
+
+- `kevm view-kcfg [save_directory] [spec_file] [--claim claim_label] ...` command takes the same basic arguments as `kevm prove --pyk ...` does, including:
+  - `save_directory` must be passed as where the KCFGs have been saved (by a previous call to `kevm prove --pyk --save-directory save_directory ...`
+  - `spec_file` is the file to look in for specifications. This file is read the same way as with `kevm prove —pyk …`; the `KProve.get_claims` is called, invoking the frontend.
+  - `--claim claim_label` option is added, but unlike `kevm prove --pyk ...`, you can only repeat it once. This option lets you select an individual claim out of the `spec_file`; if not supplied, it’s assumed that only one spec is present.
+  - `--spec-module spec_module` is also an inherited option.
```
Comment (ehildenb) [inline include/kframework/foundry.md:48]: This one is not a foundry option (neither are the code blocks following this), so should not be in this file.

File: include/kframework/foundry.md [hunk 2]
```diff
@@ -1,71 +1,86 @@
...
+- Updates the `tests/failing-symbolic.pyk` list to remove a bunch of working on RPC prover.
```
Comment (ehildenb) [inline include/kframework/foundry.md:49]: Not sure what this sentence is doing here.

File: include/kframework/foundry.md [hunk 3]
```diff
@@ -1,71 +1,86 @@
...
+
+This allows running a KEVM test specification and getting visualization working with these commands:
+
+```
+mkdir kcfgs
+kevm kompile --pyk --backend haskell tests/specs/benchmarks/verification.k --definition tests/specs/benchmarks/verification/haskell --main-module VERIFICATION --syntax-module VERIFICATION
+kevm prove tests/specs/benchmarks/address00-spec.k --definition tests/specs/benchmarks/verification/haskell --pyk --verbose --save-directory kcfgs
+kevm view-kcfg --verbose kcfgs tests/specs/benchmarks/address00-spec.k --definition tests/specs/benchmarks/verification/haskell
+```
```
Comment (ehildenb) [inline include/kframework/foundry.md:58]: These probably belong in the main README, not here.

File: README.md [hunk 4]
```diff
@@ -1,44 +1,49 @@
...
 ===========================
 
-In this repository we provide a model of the EVM in K.
+In this repository, we provide a model of the EVM in K.
 
 Fast Installation
 -----------------
 
--   `bash <(curl https://kframework.org/install)`: install [the `kup` utility](https://github.com/runtimeverification/kup)
+-   `bash <(curl https://kframework.org/install)`: install [kup package manager].
```
Comment (ehildenb) [inline README.md:9]: Why was the link removed?

```suggestion
-   `bash <(curl https://kframework.org/install)`: install [the kup package manager](https://github.com/runtimeverification/kup).
```

File: README.md [hunk 5]
```diff
@@ -80,21 +81,20 @@ make
...
             libcrypto++-dev libffi-dev libgflags-dev libjemalloc-dev libmpfr-dev       \
-            libprocps-dev libsecp256k1-dev libssl-dev libtool libyaml-dev lld-10       \
-            llvm-10-tools make maven netcat-openbsd openjdk-11-jdk pkg-config          \
-            protobuf-compiler python3 python3-dev python3-pip rapidjson-dev time      \
+            libprocps-dev libsecp256k1-dev libssl-dev libtool libyaml-dev lld-12       \
+            llvm-12-tools make maven netcat-openbsd openjdk-11-jdk pkg-config          \
+            protobuf-compiler python3 python3-dev python3-pip rapidjson-dev time       \
             zlib1g-dev
 pip3 install poetry
 ```
```
Comment (ehildenb) [inline README.md:100]: I believe we require at least Ubuntu 22 at this point? https://github.com/runtimeverification/evm-semantics/issues/1536

Comment (ehildenb) [reply README.md:100]: And ubuntu 22 comes with a newer LLVM requirement, I think llvm-14 if I'm not mistaken: https://github.com/runtimeverification/k/blob/master/.github/workflows/test-pr.yml#L92

File: README.md [hunk 6]
```diff
@@ -126,13 +126,11 @@ brew install java automake libtool gmp mpfr pkg-config maven libffi llvm@14 open
 make libsecp256k1
 ```
 
-**NOTE**: Previous versions of these instructions required the user to use either the homebrew version of `flex` or the xcode command line tools version, with the wrong option giving an error.
+**NOTE**: Previous versions of these instructions required the user to use either the homebrew version of `flex` or the XCode command line tools version, with the wrong option giving an error.
 The current recommendation is to use the homebrew version.
```
Comment (ehildenb) [inline README.md:130]: I think remove this note for now, just say "You should use the homebrew version of `flex` and XCode.

File: VERIFICATION.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
+Verification Instructions for KEVM
+==================================
+
+We assume that KEVM is installed, and the [K tutorial] has been completed.
```
Comment (ehildenb) [inline VERIFICATION.md:4]: Is this supposed to be a link? You have to provide the URL after the square-braces in parens.

File: VERIFICATION.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+This document provides instructions for kompiling and running claims using KEVM.
+
+In `tests/specs/examples`, you can find a few examples to get you started on proving with kevm.
+
+Example 1: Sum to N
+-------------------
+Have a look at the [sum-to-n-spec.k] file.
```
Comment (ehildenb) [inline VERIFICATION.md:11]:
```suggestion
Example 1: Sum to N
-------------------

Have a look at the [sum-to-n-spec.k] file.
```

File: VERIFICATION.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+It has two modules:
+  - `VERIFICATION` - containing the EVM program 
+ and a few `simplification` rules.
+  - `SUM-TO-N-SPEC` - containing the claims which will be executed.
+
```
Comment (ehildenb) [inline VERIFICATION.md:16]:
```suggestion
It has two modules:
  - `VERIFICATION` - containing the EVM program and a few `simplification` rules.
  - `SUM-TO-N-SPEC` - containing the claims which will be executed.

```

File: README.md (hunk 4 above)
Comment (anvacaru) [reply README.md:9]: It is not removed. I moved it to the bottom of the page for consistency.

File: VERIFICATION.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+Next, run the prover with:
+
+```sh
+kevm prove erc20-spec.md --backend haskell  --definition erc20-spec/haskell --pyk --claim ERC20-SPEC.decimals
+```
+Here, `--claim` tells the prover to run only the `decimals` spec from the `ERC20-SPEC` module.
+
+More to know
+------------
+To prove one of the specifications:
```
Comment (ehildenb) [inline VERIFICATION.md:74]:
```suggestion
More to know
------------

To prove one of the specifications:
```

File: VERIFICATION.md (hunk 7 above)
Comment (ehildenb) [reply VERIFICATION.md:4]: Ah, I see you put the links at the bottom of the document.

File: README.md (hunk 5 above)
Comment (anvacaru) [reply README.md:100]: I will update the Ubuntu version. I'm currently running ubuntu 22.04 with `llvm-12`.

File: VERIFICATION.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+  - `--main-module VERIFICATION`: explicitly state the main module.
+  - `--definition sum-to-n-spec/haskell`: the path where the kompiled definition is stored.
+
+Next, run the prover with:
+
+```sh
+kevm prove sum-to-n-spec.k --backend haskell --definition sum-to-n-spec/haskell
+```
+
+The expected output is `#top` which represents that all the claims have been proven.
```
Comment (ehildenb) [inline VERIFICATION.md:36]:
```suggestion
The expected output is `#Top` which represents that all the claims have been proven.
```

File: VERIFICATION.md [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+  - `VERIFICATION` - containing the EVM program and a few `simplification` rules.
+  - `SUM-TO-N-SPEC` - containing the claims which will be executed.
+
+The first step is kompiling the `.k` file with the below command.
+
+```sh
+kevm kompile sum-to-n-spec.k --pyk --backend haskell --syntax-module VERIFICATION --main-module VERIFICATION --definition sum-to-n-spec/haskell
+```
+
+In this example, the arguments used are:
```
Comment (ehildenb) [inline VERIFICATION.md:23]:
```suggestion
In this example, the arguments used are:

```

File: VERIFICATION.md [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+We assume that KEVM is installed, and the [K tutorial] has been completed.
+This document provides instructions for kompiling and running claims using KEVM.
+
+In `tests/specs/examples`, you can find a few examples to get you started on proving with kevm.
+
+Example 1: Sum to N
+-------------------
+
+Have a look at the [sum-to-n-spec.k] file.
+It has two modules:
```
Comment (ehildenb) [inline VERIFICATION.md:13]:
```suggestion
It has two modules:

```

File: VERIFICATION.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+```sh
+kevm kompile erc20-spec.md --pyk --backend haskell --syntax-module VERIFICATION --main-module VERIFICATION --definition erc20-spec/haskell
+```
+
+Next, run the prover with:
+
+```sh
+kevm prove erc20-spec.md --backend haskell  --definition erc20-spec/haskell --pyk --claim ERC20-SPEC.decimals
+```
```
Comment (ehildenb) [inline VERIFICATION.md:69]:
`````suggestion
```

`````

File: VERIFICATION.md [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+To prove one of the specifications:
+
+```sh
+kevm prove tests/specs/erc20/ds/transfer-failure-1-a-spec.k --definition tests/specs/erc20/verification/haskell
+```
+
+You can also debug proofs interactively:
+
+```sh
+kevm prove tests/specs/erc20/ds/transfer-failure-1-a-spec.k --definition tests/specs/erc20/verification/haskell --debugger
```
Comment (ehildenb) [inline VERIFICATION.md:84]:
`````suggestion
kevm prove tests/specs/erc20/ds/transfer-failure-1-a-spec.k --definition tests/specs/erc20/verification/haskell --debugger
```
`````

File: include/kframework/foundry.md [hunk 16]
```diff
@@ -1,71 +1,70 @@
...
+
+Below we are providing an example usage and a description of all the commands you can use with KEVM to improve your experience.
+
+Available Commands
+------------------
+
+Basic commands are (and each can be passed --help):
+
+- `kevm foundry-kompile`: Kompile a definition, generating claims for each Foundry test.
+The best options are:
```
Comment (ehildenb) [inline include/kframework/foundry.md:21]:
```suggestion
  The best options are:
```

File: include/kframework/foundry.md [hunk 17]
```diff
@@ -1,71 +1,70 @@
...
+   - `--regen` - needed if Solidity sources change,
+   - `-rekompile` - needed if K lemmas change, or K definition changes,
+   - `--require` - for adding an extra K file with lemmas, 
+   - `--module-import` - importing an extra K module in one of the added K files with lemmas.
+
+- `kevm foundry-prove`: Run a given proof using the KCFG-based prover (not supporting loops yet, need to fall back to typical K for that).
+The best options are:
```
Comment (ehildenb) [inline include/kframework/foundry.md:28]:
```suggestion
  The best options are:
```

File: include/kframework/foundry.md [hunk 18]
```diff
@@ -1,71 +1,70 @@
...
+   - `—reinit` - want to start over from scratch,
+   - `—no-simplify-init` - do not want to invoke simplification on all the original nodes, can be faster,
+   - `—max-depth` - increase the space between saved nodes; bigger is faster,
+   - `—max-iterations` - maximum number of nodes to store in KCFG before giving on attempting proof for that run,
+   - `—break-every-step` - save a state every opcode, slow, good for debugging,
+   - `—break-on-calls` - save a state every time a call is made, good to turn on.
+
+- `kevm foundry-show`: Display the given KCFG so far as text. 
+Options are:
```
Comment (ehildenb) [inline include/kframework/foundry.md:37]:
```suggestion
  Options are:
```

File: include/kframework/foundry.md [hunk 19]
```diff
@@ -1,71 +1,70 @@
...
+
+Available Commands
+------------------
+
+Basic commands are (and each can be passed --help):
+
+- `kevm foundry-kompile`: Kompile a definition, generating claims for each Foundry test.
+The best options are:
+   - `--regen` - needed if Solidity sources change,
+   - `-rekompile` - needed if K lemmas change, or K definition changes,
```
Comment (ehildenb) [inline include/kframework/foundry.md:23]:
```suggestion
   - `--rekompile` - needed if K lemmas change, or K definition changes,
```

File: include/kframework/foundry.md [hunk 20]
```diff
@@ -1,71 +1,70 @@
...
+   - `--module-import` - importing an extra K module in one of the added K files with lemmas.
+
+- `kevm foundry-prove`: Run a given proof using the KCFG-based prover (not supporting loops yet, need to fall back to typical K for that).
+The best options are:
+   - `—reinit` - want to start over from scratch,
+   - `—no-simplify-init` - do not want to invoke simplification on all the original nodes, can be faster,
+   - `—max-depth` - increase the space between saved nodes; bigger is faster,
+   - `—max-iterations` - maximum number of nodes to store in KCFG before giving on attempting proof for that run,
+   - `—break-every-step` - save a state every opcode, slow, good for debugging,
+   - `—break-on-calls` - save a state every time a call is made, good to turn on.
```
Comment (ehildenb) [inline include/kframework/foundry.md:34]:
```suggestion
   - `—break-on-calls` - save a state every time a call is made, good to turn on.
   - `--verbose` - output what the prover is doing to make sure it's making progress.
```

File: README.md [hunk 21]
```diff
@@ -47,29 +52,25 @@ These files are used for testing the semantics itself:
...
 #### Installing Z3
 
-KEVM requires Z3 version 4.8.15, which you may need to install from a source
-build if your package manager supplies a different version. To do so, follow the
-instructions
-[here](https://github.com/Z3Prover/z3#building-z3-using-make-and-gccclang) after
-checking out the correct tag in the Z3 repository:
-
+KEVM requires Z3 version 4.8.15, which you may need to install from a source build if your package manager supplies a different version.
+To do so, follow the instructions [here](https://github.com/Z3Prover/z3#building-z3-using-make-and-gccclang) after checking out the correct tag in the Z3 repository:
```
Comment (ehildenb) [inline README.md:73]:
```suggestion
To do so, follow the instructions [here](https://github.com/Z3Prover/z3#building-z3-using-make-and-gccclang) after checking out the correct tag in the Z3 repository:

```

File: README.md [hunk 22]
```diff
@@ -247,25 +253,16 @@ Run the file `tests/ethereum-tests/LegacyTests/Constantinople/VMTests/vmArithmet
...
 
-To run proofs, you can similarly use `kevm`.
-For example, to prove one of the specifications:
-
-```sh
-kevm prove tests/specs/erc20/ds/transfer-failure-1-a-spec.k --verif-module VERIFICATION
-```
-
-You can also debug proofs interactively:
+### Keeping in mind while developing
```
Comment (ehildenb) [inline README.md:256]:
```suggestion
### Keeping in mind while developing

```

File: README.md [hunk 23]
```diff
@@ -275,91 +272,82 @@ To set up nix flakes you will need to be on `nix` 2.4 or higher and follow the i
 For example, if you are on a standard Linux distribution, such as Ubuntu, first [install nix](https://nixos.org/download.html#download-nix)
 and then enable flakes by editing either `~/.config/nix/nix.conf` or `/etc/nix/nix.conf` and adding:
 
-```
+```sh
```
Comment (ehildenb) [inline README.md:275]:
`````suggestion
```
`````

File: README.md [hunk 24]
```diff
@@ -275,91 +272,82 @@ To set up nix flakes you will need to be on `nix` 2.4 or higher and follow the i
...
-By default, Nix will build the project and its transitive dependencies from
-source, which can take up to an hour. We recommend setting up
-[the binary cache](https://app.cachix.org/cache/kore) to speed up the build
-process significantly. You will also need to add the following sections to `/etc/nix/nix.conf` or, if you are a trusted user, `~/.config/nix/nix.conf` (if you don't know what a "trusted user" is, you probably want to do the former):
+By default, Nix will build the project and its transitive dependencies from source, which can take up to an hour.
+We recommend setting up [the binary cache](https://app.cachix.org/cache/kore) to speed up the build process significantly.
+You will also need to add the following sections to `/etc/nix/nix.conf` or, if you are a trusted user, `~/.config/nix/nix.conf` (if you don't know what a "trusted user" is, you probably want to do the former):
 
-```
+```sh
```
Comment (ehildenb) [inline README.md:286]:
`````suggestion
```
`````

File: README.md [hunk 25]
```diff
@@ -275,91 +272,82 @@ To set up nix flakes you will need to be on `nix` 2.4 or higher and follow the i
...
 trusted-public-keys = ... hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ=
 substituters = ... https://cache.iog.io
 ```
 
 i.e. if the file was originally
 
-```
+```sh
```
Comment (ehildenb) [inline README.md:293]:
`````suggestion
```
`````

File: README.md [hunk 26]
```diff
@@ -275,91 +272,82 @@ To set up nix flakes you will need to be on `nix` 2.4 or higher and follow the i
...
 substituters = https://cache.nixos.org
 trusted-public-keys = cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
 ```
 
 it will now read
 
-```
+```sh
```
Comment (ehildenb) [inline README.md:300]:
`````suggestion
```
`````

**Review by ehildenb — APPROVED** [review]

**Pull request merged**